Notes/Algorithms/Data Structures/Hash-Tables.c

```diff
@@ -1,230 +1,244 @@
-        printf("The key %d was successfully found as expected. \n", key);
+        printf("The key/value pair: %s and %d was successfully found as expected. \n", key, value);
         return SUCCESS;
     } else if (success == ERROR && expectedOutcome == ERROR) {
-        printf("The key %d was not found as expected. \n", key);
+        printf("The key/value pair: %s and %d was not found as expected. \n", key, value);
         return SUCCESS;
     }
-    printf("Mismatch: An unknown error has occurred. \n");
+    printf("Mismatch has occurred. Result does not match expected outcome for success or failure. \n");
     return ERROR;
 }
 
+// buildSearchTests
+// creates 5 test cases expected to pass the testSearch function
+// creates 5 test cases (out of range) expected to fail the testSearch function
 void buildSearchTests() {
-printf("Build Search Test Passes: \n");
-    testSearch(1, SUCCESS);
-    testSearch(8, SUCCESS);
-    testSearch(3, SUCCESS);
-    testSearch(4, SUCCESS);
-    testSearch(9, SUCCESS);
+    printf("Build Search Test Passes: \n");
+    testSearch("KeyTwo", 20, SUCCESS);
+    testSearch("KeyEight", 80, SUCCESS);
+    testSearch("KeyThree", 30, SUCCESS);
+    testSearch("KeyFour", 40, SUCCESS);
+    testSearch("KeyNine", 90, SUCCESS);
 
     //5 Fail
     printf("Build Search Test Fails: \n");
-    testSearch(99, ERROR);
-    testSearch(-9, ERROR);
-    testSearch(11, ERROR);
-    testSearch(20, ERROR);
-    testSearch(-20, ERROR);
+    testSearch("KeyNinety", 0, ERROR);
+    testSearch("LaLa", 10, ERROR);
+    testSearch("Error", 20, ERROR);
+    testSearch("-KeyNine", 30, ERROR);
+    testSearch("Bad", 40, ERROR);
 }
-
-//To-do: Like above need to make adjustments based on potential insertion collisions messing up the hashing function
-// Also saw an alternative way to do delete is just to set the key to -1?
-// deleteItem takes in an Item struct
-// finds the key of the provided Item then passes it to the helper function to hash it and find the corresponding index
-// if the key at the position in the hashTable indicated by the hashing function matches the received key then sets an empty delete Item equal to the Item at the hashIndex
-// if a failure occurs return an ERROR code
-void deleteItem(int key, int * success) {
-    int hashIndex = hashing(key);
-    if (testHash[hashIndex]->key == key) {
-        free(testHash[hashIndex]);
-        insert(9, key, NULL, *success);
-        *success = SUCCESS;
+ 
+// deleteItem
+// if key/value pair is successfully deleted return SUCCESS code
+// if key/value pair cannot be deleted return ERROR code 
+int deleteItem(HashTable * hashTable, char key[], int value) {
+    int hashIndex = hashing(hashTable, key);
+    for (int i = 0; i < hashTable->size; i++) {
+        int index = (hashIndex + i) % hashTable->size;
+        if (hashTable->items[index]->value == value && strcmp(hashTable->items[index]->key, key) == 0) {
+            free(hashTable->items[index]);
```

Comment:
In `createItem` you `strdup`.  Given the cleanup logic here, I believe that memory (for the duplicated string) is leaking.  I'd recommend a counterpart `deleteItem` that does the inverse/cleanup corresponding to `createItem`.

Reply:
Can you expand on this - I thought free would also be freeing the index->value and index->key? Or are you saying that before doing free(hashTable->items[index]) there should be individual frees for value + key?

Reply:
Updated per conversation about freeing pointers before Item [commit c79576986c842dfcc20394b1317fb3285ba708b7](https://github.com/kcarrel/Training-Notes/pull/12/commits/c79576986c842dfcc20394b1317fb3285ba708b7)